modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java

```diff
@@ -265,88 +270,135 @@ protected int processStartTag() throws Exception {
+		int totalPages = (int)Math.ceil(
+			(double)_totalItems / (double)_activeDelta);
+
+		if (_ellipsisBuffer != 0) {
+			int activeIndex = _activePage - 1;
+
+			int[] pageNumberItems = new int[totalPages];
+
+			for (int i = 0; i < totalPages; i++) {
+				pageNumberItems[i] = i + 1;
+			}
+
+			int firstItem = pageNumberItems[0];
 
-		int ellipsisCount = _activePage + _ellipsisBuffer;
+			_processItem(pageContext, firstItem);
 
-		for (int i = 1; i < ellipsisCount; i++) {
-			jspWriter.write("<li class=\"page-item");
+			int leftIndex = activeIndex - _ellipsisBuffer;
 
-			if (i == _activePage) {
-				jspWriter.write(" active");
+			int[] leftItems = Arrays.copyOfRange(
+				pageNumberItems, 1, Math.max(leftIndex, 1));
+
+			if (leftItems.length > 1) {
+				_processEllispis(pageContext);
+			}
+			else {
+				int leftItem = leftItems[0];
+
+				_processItem(pageContext, leftItem);
 			}
 
-			if ((_disabledPages != null) &&
-				_disabledPages.contains(Integer.valueOf(i))) {
+			int lastIndex = pageNumberItems.length - 1;
+
+			int[] centerItems = Arrays.copyOfRange(
+				pageNumberItems, Math.max(activeIndex - _ellipsisBuffer, 1),
+				Math.min(activeIndex + _ellipsisBuffer + 1, lastIndex));
 
-				jspWriter.write(" disabled");
+			for (int centerItem : centerItems) {
+				_processItem(pageContext, centerItem);
 			}
 
-			jspWriter.write("\">");
+			int rightIndex = activeIndex + _ellipsisBuffer + 1;
 
-			buttonTag = new ButtonTag();
+			if (rightIndex > lastIndex) {
+				rightIndex = lastIndex;
+			}
 
-			buttonTag.setCssClass("page-link");
-			buttonTag.setDisplayType("unstyled");
-			buttonTag.setLabel(String.valueOf(i));
+			int[] rightItems = Arrays.copyOfRange(
+				pageNumberItems, rightIndex, Math.max(lastIndex, rightIndex));
 
-			buttonTag.doTag(pageContext);
+			if (rightItems.length > 1) {
+				_processEllispis(pageContext);
+			}
+			else if (rightItems.length == 1) {
+				int rightItem = rightItems[0];
 
-			jspWriter.write("</li>");
+				_processItem(pageContext, rightItem);
+			}
+
+			if (totalPages > 1) {
+				int lastItem = pageNumberItems[lastIndex];
+
+				_processItem(pageContext, lastItem);
+			}
+		}
+		else {
+			for (int i = 1; i < (totalPages + 1); i++) {
+				_processItem(pageContext, i);
+			}
 		}
 
-		jspWriter.write("<li class=\"dropdown page-item\">");
+		jspWriter.write("<li class=\"page-item\">");
 
 		buttonTag = new ButtonTag();
 
-		buttonTag.setCssClass("dropdown-toggle page-link");
+		buttonTag.setCssClass("page-link");
 		buttonTag.setDisplayType("unstyled");
-		buttonTag.setLabel("...");
+		buttonTag.setIcon("angle-right");
 
 		buttonTag.doTag(pageContext);
 
-		jspWriter.write("</li>");
+		jspWriter.write("</li></ul>");
 
-		for (int i = ellipsisCount + 1; i < (totalPages + 1); i++) {
-			jspWriter.write("<li class=\"page-item");
+		return SKIP_BODY;
+	}
 
-			if (i == _activePage) {
-				jspWriter.write(" active");
-			}
+	private void _processEllispis(PageContext pageContext) throws Exception {
```

Comment:
There's a small typo here, was this meant to be `_processEllipsis`? see the `sp` transposition.

Reply:
Thanks for the catch @jbalsas :facepalm: I always type ellipsis wrong